prometheus-ksonnet/lib/alertmanager.libsonnet

```diff
@@ -2,17 +2,9 @@
   local replicas = self._config.alertmanager_cluster_self.replicas,
   local isGlobal = self._config.alertmanager_cluster_self.global,
   local isGossiping = replicas > 1 || isGlobal,
-  local peers = if isGlobal
-  then
+  local peers = if isGossiping then
     [
-      'alertmanager-%d.alertmanager.%s.svc.%s.%s:%s' % [i, $._config.namespace, cluster, $._config.cluster_dns_tld, $._config.alertmanager_gossip_port]
-      for cluster in std.objectFields($._config.alertmanager_clusters)
-      if $._config.alertmanager_clusters[cluster].global
-      for i in std.range(0, $._config.alertmanager_clusters[cluster].replicas - 1)
-    ]
-  else if isGossiping then
-    [
-      'alertmanager-%d.alertmanager.%s.svc.%s.%s:%s' % [i, $._config.namespace, $._config.cluster_name, $._config.cluster_dns_tld, $._config.alertmanager_gossip_port]
+      'alertmanager-%d.alertmanager-gossip.%s.svc.%s:%s' % [i, $._config.namespace, $._config.cluster_dns_suffix, $._config.alertmanager_gossip_port]
```

Comment:
This list comprehension looks like it won't iterate anymore over Alertmanagers in different clusters. But that's the whole idea of an HA Alertmanager cluster: Have instances in different zones, together creating a global AM cluster.

Reply:
I think that incorporating multiple clusters in an HA arrangement is a good goal. 
However, it is not clear to me how the previous code would  achieve the intended result. Unless I'm missing something, it was attempting to use the `internal DNS` designation of each cluster, e.g. `cluster.local` as the differentiator of each cluster.  Maybe Grafana uses some mechanism for exposing the internal DNS tld for each of your federated clusters?  In my experience, the internal DNS identifiers in each cluster are private to the cluster, and different mechanisms are used for aggregating services across clusters.   Am I just getting the intent of that code wrong?

Reply:
Maybe. I'll have to discuss this with my infrastructure team mates.

To clarify: The code here was written specifically to run a cross-colo AM cluster, and said cluster has been running at Grafana with this code for quite some time.

Reply:
Hey @rrichardson

Beorn is right, we have CoreDNS configured that passes DNS queries to other clusters. I just merged https://github.com/grafana/jsonnet-libs/pull/472 that tries to create more manageable building blocks for a highly available alertmanager setup, please have a look if that can help you.

Maybe we can come up with a mixin for the service-based? For example a `withGossipService` function that modifies the alertmanager setup to match your use case? Then it doesn't break existing setups.